### ISSUE TITLE
对于对比损失函数的一个问题？

### ISSUE THREAD

**Issue by dugzzuli** [Guanzhou-Ke/conan#3]
`

class SimCLRLoss(BaseLoss, ABC):

    def __init__(self, args):
        super().__init__()
        self.device = args.device
        self.temp = args.temperature

    @staticmethod
    def _norm(mat):
        return torch.nn.functional.normalize(mat, p=2, dim=1)

    @classmethod
    def _normalized_projections(cls, ps):
        n = ps.size(0) // 2
        h1, h2 = ps[:n], ps[n:] # 对比损失，不是相当于两个视角的样本进行对比吗？而这个是去了每个视角的前50个，和后50个。
        h2 = cls._norm(h2)
        h1 = cls._norm(h1)
        return n, h1, h2

    def _loss_func(self, ps): 
       #经过调试按照默认参数，ps的维度为100*200-> ps = torch.cat([cp, p], dim=-1)
       #为什么是通过第二个维度串联，不应该是通过第一个维度串联吗？ ps的维度应该表示为200X100

        n, h1, h2 = self._normalized_projections(ps)

        labels = torch.arange(0, n, device=self.device, dtype=torch.long)
        masks = torch.eye(n, device=self.device)
    
        logits_aa = ((h1 @ h1.t()) / self.temp) - masks * self.large_num #相当于前50个样本进行计算，
        logits_bb = ((h2 @ h2.t()) / self.temp) - masks * self.large_num# 相当于后50个样本进行计算

        logits_ab = (h1 @ h2.t()) / self.temp
        logits_ba = (h2 @ h1.t()) / self.temp

        loss_a = torch.nn.functional.cross_entropy(torch.cat((logits_ab, logits_aa), dim=1), labels) # 不太理解这个步骤
        loss_b = torch.nn.functional.cross_entropy(torch.cat((logits_ba, logits_bb), dim=1), labels)

        loss = (loss_a + loss_b)

        return loss

    def __call__(self, ps):
        return self._loss_func(ps)
`

对这个损失函数，我有一些问题。

**Comment by Guanzhou-Ke** [comment]
您好，非常感谢您关注我们的工作！

将一致性表征和视图表征拼接是为了计算对比相似度，第二个维度才是feature特征，第一个维度是batch。

simclr的损失在我们这里被改造成了视图间的对比，而不是增强样本的对比。

您可以理解size(0)是取对半的batch做正负样本

**Comment by dugzzuli** [comment]
非常感谢您的解答
不过我没有理解到这个视图之间的对比，是通过那一句代码？怎么对比的？
ps通过第二个维度进行串联，shape应该表示为100X200
h1的shape为50*200.
“(h1 @ h1.t())” 这个计算得到的是对半的batch做正负样本，也就是说假设batch为100，那么取前50个和后50个进行对比，为什么这么做呢，不应该是一致性表示的100个样本和视角表示100个样本之间的对比吗？
有点不太理解，
